src/Compression/CompressionCodecEncrypted.h

```diff
@@ -79,16 +85,19 @@ namespace DB
 
     private:
         static std::string lastErrorString();
-        static std::string deriveKey(const std::string_view & master_key);
         static void encrypt(const std::string_view & plaintext, char * ciphertext_and_tag);
-        static void decrypt(const std::string_view & ciphertext_and_tag, char * plaintext);
+        static void decrypt(const std::string_view & ciphertext_and_tag, char * plaintext, UInt64 key_id=current_key_id);
+
+
+        static std::unordered_map<UInt64, String> keys_storage;
+        static UInt64 current_key_id;
 
         /** A private class that holds keys derived from the master
           * key.
           */
         struct KeyHolder : private boost::noncopyable
         {
-            KeyHolder(const std::string_view & master_key);
+            explicit KeyHolder(const std::string_view & master_key);
             ~KeyHolder();
 
             std::string keygen_key;
```

Comment:
`keygen_key` is not used anymore, so we can remove it.
Also struct `KeyHolder` looks unnecessary with only `ctx` contained in it.
So it seems we can remove the struct too and use the cipher instead.